packages/engine/src/scene/systems/InstancingSystem.tsx

```diff
@@ -0,0 +1,157 @@
+/*
+CPAL-1.0 License
+
+The contents of this file are subject to the Common Public Attribution License
+Version 1.0. (the "License"); you may not use this file except in compliance
+with the License. You may obtain a copy of the License at
+https://github.com/ir-engine/ir-engine/blob/dev/LICENSE.
+The License is based on the Mozilla Public License Version 1.1, but Sections 14
+and 15 have been added to cover use of software over a computer network and
+provide for limited attribution for the Original Developer. In addition,
+Exhibit A has been modified to be consistent with Exhibit B.
+
+Software distributed under the License is distributed on an "AS IS" basis,
+WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the
+specific language governing rights and limitations under the License.
+
+The Original Code is Infinite Reality Engine.
+
+The Original Developer is the Initial Developer. The Initial Developer of the
+Original Code is the Infinite Reality Engine team.
+
+All portions of the code written by the Infinite Reality Engine team are Copyright © 2021-2025
+Infinite Reality Engine. All Rights Reserved.
+*/
+
+import {
+  defineSystem,
+  Entity,
+  EntityTreeComponent,
+  EntityUUIDPair,
+  getComponent,
+  PresentationSystemGroup,
+  removeComponent,
+  setComponent,
+  useComponent,
+  useQuery,
+  useQueryBySource,
+  UUIDComponent
+} from '@ir-engine/ecs'
+import { TransformComponent } from '@ir-engine/spatial'
+import { NameComponent } from '@ir-engine/spatial/src/common/NameComponent'
+import { RigidBodyComponent } from '@ir-engine/spatial/src/physics/components/RigidBodyComponent'
+import { MeshComponent } from '@ir-engine/spatial/src/renderer/components/MeshComponent'
+import { VisibleComponent } from '@ir-engine/spatial/src/renderer/components/VisibleComponent'
+import React, { useEffect } from 'react'
+import { InstancedBufferAttribute, InstancedMesh, Material } from 'three'
+import { GLTFComponent } from '../../gltf/GLTFComponent'
+import { InstancingComponent } from '../components/InstancingComponent'
+import { VariantComponent } from '../components/VariantComponent'
+import { useEntity } from '../functions/useEntity'
+
+interface Level {
+  src: string
+  metadata: Record<string, any>
+}
+
+export const InstancingSystem = defineSystem({
```

Comment:
This system is making some assumptions about what instances and variants will be used for. We have plans to refactor these that will achieve the same thing without making assumptions.